src/System.Data.SqlClient/src/System/Data/SqlClient/SNI/SNIProxy.cs

```diff
@@ -698,4 +576,260 @@ public SNIError GetLastError()
             return SNILoadHandle.SingletonInstance.LastError;
         }
     }
+
+    internal class DataSource
+    {
+
+        private const char CommaSeparator = ',';
+        private const char BackSlashSeparator = '\\';
+        private const char ForwardSlashSeparator = '/';
+        private const string DefaultHostName = "localhost";
+        private const string DefaultSqlServerInstanceName = "mssqlserver";
+        private const string PipeBeginning = @"\\";
+        private const string PipeToken = "pipe";
+
+        internal enum Protocol { TCP, NP, None, Admin };
+
+        internal Protocol ConnectionProtocol = Protocol.None;
+
+        internal string ServerName { get; private set; }
+        internal int Port { get; private set; } = -1;
+
+        public string InstanceName { get; internal set; }
+
+        public string PipeName { get; internal set; }
+
+
+        private string _workingDataSource;
+        private string _dataSourceAfterTrimmingProtocol;
+        internal bool IsBadDataSource { get; private set; } = false;
+
+        private DataSource(string dataSource)
+        {
+            // Remove all whitespaces from the datasource and all operations will happen on lower case.
+            _workingDataSource = dataSource.Trim().ToLower();
+
+            int firstIndexOfColon = _workingDataSource.IndexOf(':');
+
+            PopulateProtocol();
+
+            _dataSourceAfterTrimmingProtocol = (firstIndexOfColon >= -1) && ConnectionProtocol != DataSource.Protocol.None
```

Comment:
The check effectively translates to true. By the time the code gets here, either there is a protocol or there is none. I don't need to check the firstIndexOfColon, because the ConnectionProtocol is a result of checking it. 
If ConnectionProtocol is None, then there was no ":" present or it was a IP v6 address. 
We still need to SubString if we found a protocol, to get the datasource without the protocol.